### PR TITLE
Reuse the DominatorTree postorder travesal in BlockLoweringOrder

### PR DESCRIPTION
**Pull request by elliottt**
Rework `BlockLoweringOrder` construction to use the postorder traversal created by the `DominatorTree`. Additionally, construct fewer critical edge splits by identifying edges up-front that will need splitting, and otherwise allowing regalloc2 to place moves on either the entry or exit edges of a block. Identifying critical edges up-front often yields fewer critical edge splits than the previous algorithm, leading to lowering orders with fewer blocks overall. This change leads to small but measurable improvements in sightglass benchmarks for compilation on both `bz2` and `spidermonkey` wasm.

The meat of this change is in the first commit, which both changes the block lowering order implementation and propagates a heuristic from the old implementation's postorder construction to the DominatorTree's implementation. This heuristic (reversing the children before pushing them on the work stack) is responsible for much of the test change fallout in this PR, but is important in that without it the new lowering order sees significantly more cache misses on the bz2 benchmark.

### CODE REVIEW

**Comment by cfallin** [comment]
I'll review this in detail tomorrow, but one thing I think would be good to explore / understand (if only because I sometimes find that fully explaining why a problem occurred leads to uncovering other surprises, or bugs, or ... basically, [Chesterton's Fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence)): what cases did the existing code not cover that this does?

I'm somewhat surprised because the existing implementation already was supposed to be creating lowered blocks only when necessary: its `Orig` / `Edge` / `EdgeAndOrig` / `OrigAndEdge` descriptors allow for merging the edge-block into either pred or succ, and for any non-critical edge (such that either pred or succ has at most one succ or pred, respectively), this merging should be possible. In other words, the problem description here "now we only create edge blocks for critical edges" is surprising to me because we were supposed to already be doing this. Was there a case missing or overly conservative, for example? Or is the new implementation not accounting for something?

I see that tests are failing with RA2 rejecting some lowered code ("disallowed branch arg", which occurs when the CFG forces edge moves to the pred side of an edge but the pred ends in a conditional: in this case, we have nowhere to put the moves for just one side of the conditional), which IMHO means we should be extra-careful in understanding exactly what the difference is and what new cases we are now allowing!

**Comment by elliottt** [comment]
The RA2 error was from treating `br_table` with an empty jump table as an unconditional jump. I ported forward the workaround from the previous implementation, which was to mark blocks that end in `br_table` as having at least 2 outgoing edges.

**Comment by elliottt** [comment]
There were a few different motivations for this PR, that came out of discussions with @jameysharp, and recent refactorings that get us to a state where we can now assume all branch instructions terminate basic blocks.

Starting with the latter first, I was idly investigating what it would take to have the output of `BlockLoweringOrder::succ_indices` return a single instruction along with a slice of the indices it branches to. This would allow simplifications in a few other places where we jump through hoops to assert that they are all the same instruction now. This was a pretty daunting refactoring with the previous implementation, as the `LoweredBlock` type makes use of the branch instruction in many places, and restructuring metadata used during the construction of the postorder traversal resisted multiple attempts by both me and @jameysharp.

As to the former, @jameysharp and I had been discussing how it would be nice to reuse the postorder  traversal that's cached by the `DominatorTree` to drive the block lowering order. This is nice for a few reasons: it's not unnecessarily recomputing the order which while cheap to compute is not nothing; it allows us to specify explicitly that we only split critical edges, which was easier for me to think about than the state machine in the current implementation.

Additionally this implementation allows us to avoid splitting edges for unconditional jumps leaving one block but targeting another that has multiple inputs, which in some cases does produce lowering orders with fewer critical edge nodes. One concrete case is the diamond from the test suite: an entry with a `brif` whose target blocks both branch to the same exit node. Previously that would introduce two edge blocks on the unconditional branches to the exit block, but with this implementation those introduced nodes are not necessary. Investigating how the `LoweredBlock` values are used by the RA2 interface is what led @jameysharp and I to believe that this would be a reasonable change to make, as it treats all `LoweredBlock` cases the same unless they are the pure `Edge` case.

**Comment by cfallin** [comment]
> One concrete case is the diamond from the test suite: an entry with a brif whose target blocks both branch to the same exit node. Previously that would introduce two edge blocks on the unconditional branches to the exit block, but with this implementation those introduced nodes are not necessary.

Hmm... so, as silly as it may seem in the minimal example, in general we *do* actually need separate edge blocks here, because of blockparams. Specifically, imagine that each edge from A to B assigned a different list of values to the blockparams in B: then RA2 needs a location to do that permutation for each edge. This is why the definition of edge-criticality in the existing code has to do with number of successors and predecessors, without collapsing.

**Comment by elliottt** [comment]
> > One concrete case is the diamond from the test suite: an entry with a brif whose target blocks both branch to the same exit node. Previously that would introduce two edge blocks on the unconditional branches to the exit block, but with this implementation those introduced nodes are not necessary.
> 
> Hmm... so, as silly as it may seem in the minimal example, in general we _do_ actually need separate edge blocks here, because of blockparams. Specifically, imagine that each edge from A to B assigned a different list of values to the blockparams in B: then RA2 needs a location to do that permutation for each edge. This is why the definition of edge-criticality in the existing code has to do with number of successors and predecessors, without collapsing.

I think that's still easier to implement in the context of these changes, as we're currently adding an edge block only for critical edges; it would be really easy to rephrase this to add an edge block when there's a critical edge, or the successor block has multiple predecessors.

**Comment by cfallin** [comment]
> I think that's still easier to implement in the context of these changes, as we're currently adding an edge block only for critical edges; it would be really easy to rephrase this to add an edge block when there's a critical edge, or the successor block has multiple predecessors.

That's possibly equivalent to the current code, but phrased in a (to me) more confusing way. Given a CFG with two edges from A to B, these edges *are* critical edges; no need to distinguish the special case with multiple preds. Strictly speaking I guess one needs to see the CFG as a hypergraph (every edge has an identity) -- and this is why the edges use branch instruction IDs to disambiguate.

That suggests a different fix to me: why not start with the current code, but use successor-index, rather than branch instruction, to disambiguate edges?

Given the above fix, I suspect this PR would produce exactly equivalent code to the current code. I'm interested to see whether this is actually the case though.

**Comment by jameysharp** [comment]
I'd really like to drop the "Treat br_table with an empty jump table as multiple block exits" workaround. The problem it's dealing with is that the input to such instructions is a register use that RA2 would have to account for when emitting moves between blocks. (I didn't understand that when looking at this with Trevor last week.) However in the only case where this can occur, that use is effectively dead: it can't have any effect on the execution of the program. I don't know if we should just make it illegal to have a `br_table` with an empty jump table (you can always replace it with a `jump`), or do a legalization pass at some point, but I don't like kludging the block-lowering order as a "solution".

Regarding the edge-splits that this PR removes compared to the existing implementation: The current implementation supports fusing `EdgeAndOrig` or `OrigAndEdge` pairs, but if both fusions are legal it picks only one instead of emitting something like `EdgeAndOrigAndEdge`. That's the case this PR improves on. This is mentioned in the module comment:

```
//! (note that the edges into CLIF blocks 1 and 2 could be merged with those
//! blocks' original bodies, but the out-edges could not because for simplicity
//! in the successor-function definition, we only ever merge an edge onto one
//! side of an original CLIF block.)
```

The key observation, I think, is that RA2 can't actually distinguish between `Orig`, `EdgeOrig`, and `OrigEdge` blocks. The only two kinds of blocks we hand it in [machinst/lower.rs](https://github.com/bytecodealliance/wasmtime/blob/d711872d6367d2f50a14461c48ae4f19d571f984/cranelift/codegen/src/machinst/lower.rs#L1009-L1063) are either pure critical-edge splitting blocks without any original instructions in them, or original blocks where RA2 may safely add moves to the beginning and end as needed. This PR makes that distinction explicit.

I think you two are talking past each other regarding which edges are critical.

I think the case Chris is talking about is if there are only two blocks (A and B), and A has two edges to B. In that case, yes, both of those edges are critical: they're both edges from a block with multiple successors and to a block with multiple predecessors. As you pointed out, the two edges may have different block parameters so we can't in general place the appropriate moves in either A or B. Those predecessors and successors aren't distinct if we look only at block IDs, but they are distinct edges, and that's what this PR looks at.

The case which Trevor is talking about is when there are four blocks (A, B, C, D), with a conditional branch from A to either B or C, and then B and C both unconditionally branch to D. In that case there are no critical edges, even with block parameters. All of the moves can be attached to block B or C, because they each have exactly one predecessor and one successor. The existing implementation inserts unnecessary edge splits in this case.

I believe this PR already captures the definition that you intended, Chris. It's much simpler than the existing implementation, in my opinion, because it only needs to iterate over all the edges and decide whether each edge is critical or not. That decision is simple after counting all the predecessors and successors of each block.

The only tricky bit in this PR, I think, is in deciding where to insert the edge-splitting blocks in the pre-computed RPO. This PR places them immediately after their predecessors, if I remember correctly, but several other choices would be equally reasonable, I think. There's no "good" choice, as far as I can figure, but this is one choice that I think still produces a valid RPO, which is a reasonable heuristic.

File: cranelift/codegen/src/machinst/blockorder.rs [hunk 1]
```diff
@@ -110,390 +92,196 @@ pub struct BlockLoweringOrder {
...
-        edge_inst: Inst,
-        /// The successor index in this edge, to distinguish multiple
-        /// edges between the same block pair.
-        succ_idx: usize,
-        /// The successor CLIF block.
+
+        /// The successor block.
         succ: Block,
+
+        /// The branch index
```
Comment (cfallin) [inline cranelift/codegen/src/machinst/blockorder.rs:111]: Is this the index of the branch, or the index of the target within the single branch that can exist? (I prefer the existing `succ_idx` name for that, if that's what it is)

File: cranelift/codegen/src/machinst/blockorder.rs [hunk 2]
```diff
@@ -110,390 +92,196 @@ pub struct BlockLoweringOrder {
...
-        for block in f.layout.blocks() {
-            if let Some(inst) = f.layout.last_inst(block) {
-                // If the block has a branch with any "fixed args"
-                // (not blockparam args) ...
-                if f.dfg.insts[inst].opcode().is_branch() && f.dfg.inst_fixed_args(inst).len() > 0 {
-                    // ... then force a minimum successor count of
-                    // two, so the below algorithm cannot put
-                    // edge-moves on the end of the block.
-                    block_out_count[block] = std::cmp::max(2, block_out_count[block]);
+                if Opcode::BrTable == f.dfg.insts[inst].opcode() {
```
Comment (cfallin) [inline cranelift/codegen/src/machinst/blockorder.rs:177]: This isn't quite the same as the existing logic, though today it evaluates the same: the existing logic is that if there are any fixed args (ie, non-blockparam args), we need to ensure no edge-moves to succs end up in this block. Branch tables with empty tables can result in this, but I'd prefer to encode the actual restriction for robustness (in other words, I want correctness written explicitly rather than an emergent property). Can we take the `f.dfg.insts[inst].opcode().is_branch() && f.dfg.inst_fixed_args(inst).len() > 0` condition and use it here?

**Review by cfallin — APPROVED** [review]
Ah, that makes sense -- thanks for your patience with me! I finally understand (what was admittedly in the comments the whole time): the improvement in code quality comes from merging pre- and post-edge blocks onto the same original block.

Given that, I agree this is an improvement and is well-motivated. Some comments below -- the main one having to do with the implicit-out-edges when branches have parameters. Regarding legalization beforehand vs. having this bit of logic here: I actually think this is a bit more coherent, given the mindset that we are encoding knowledge of *edge-move placement* (not just critical-edge placement) here. Requiring legalization beforehand or requiring `br_table` not to have an empty table or something of that sort is going a level or two more indirect, imposing ad-hoc requirements on the structure of the CFG itself or properties of branch instructions. It seems least bad to me to say "a data input to a branch precludes edge moves" in one place, but I'm open to better ways of writing this still...

File: cranelift/codegen/src/machinst/blockorder.rs (hunk 2 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/blockorder.rs:177]: The third option, just to record it for completeness here, is to make RA2 handle this edge (pun intended) case properly, by considering a branch's uses to come after edge-moves if they're inserted before the branch. I'm not advocating for that by any means, but I just wanted to note that end of the spectrum exists and is, in some manner of thinking, the cleanest :-)

Thanks for writing out the reasoning here, in more depth than I had considered -- given that we have the backstop of the checks in RA2 itself, running during fuzzing, I'm fine with this as-written for now. So, LGTM with just the above `succ_idx` change.
